# 4.6 Combined Pickle Dataframe Cont.- Task Steps 3-9

# Importing Libraries

In [1]:
# Import Libraries
import pandas as pd
import numpy as np
import os

# Import Dataframes

In [6]:
# Import data set
path = r'/Users/jordanhenghold/Documents/06-09-2023 Instacart Basket Analysis'

In [7]:
# Import df_prods
df_prods =  pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'prods_checked.csv'), index_col = False)

# Step 3 - Import Pickle File

In [9]:
# Import orders_products_combined
df_ords_prods_combined = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_combined.pkl'))

# Step 4 - Check the Shape of Dataframe

In [10]:
df_ords_prods_combined.shape

(32434489, 13)

Once my tutor responds to my note this should read "(32434489, 12)" because the eval_set column has yet to be permanently deleted even though I have taken the proper precautions, so I need help with that! This proponent has yet to skew my data for misunderstandings so it is fine until then. :)

# Step 5 - Merge Dataframe with Prods Data

In [11]:
# Check the columns of df_ords_prods_combined
df_ords_prods_combined.head()

Unnamed: 0.1  Unnamed: 0  order_id  user_id eval_set  order_number  \
0             0           0   2539329        1    prior             1   
1             0           0   2539329        1    prior             1   
2             0           0   2539329        1    prior             1   
3             0           0   2539329        1    prior             1   
4             0           0   2539329        1    prior             1   

   orders_day_of_week  order_hour_of_day  days_following_previous_order  \
0                   2                  8                            NaN   
1                   2                  8                            NaN   
2                   2                  8                            NaN   
3                   2                  8                            NaN   
4                   2                  8                            NaN   

   product_id  add_to_cart_order  reordered _merge  
0         196                  1          0   both  
1       14084                  2          0   both  
2       12427                  3          0   both  
3       26088                  4          0   both  
4       26405                  5          0   both

*** Must delete eval_set column permanently

In [12]:
# Check df_prods shape
df_prods.shape

(49693, 6)

In [13]:
# Check df_prods columns
df_prods.head()

Unnamed: 0  product_id                                       product_name  \
0           0           1                         Chocolate Sandwich Cookies   
1           1           2                                   All-Seasons Salt   
2           2           3               Robust Golden Unsweetened Oolong Tea   
3           3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...   
4           4           5                          Green Chile Anytime Sauce   

   aisle_id  department_id  prices  
0        61             19     5.8  
1       104             13     9.3  
2        94              7     4.5  
3        38              1    10.5  
4         5             13     4.3

In [14]:
# Drop existing merge column (drop b/c error in following code to merge)
df_ords_prods_combined = df_ords_prods_combined.drop(columns = ['_merge'])

In [15]:
# Product_id matches in both dataframes - use as the key to merge
df_merged = df_ords_prods_combined.merge(df_prods, on = 'product_id', indicator = True)

In [20]:
df_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_following_previous_order  product_id  add_to_cart_order  reordered  \
0                            NaN         196                  1          0   
1                           15.0         196                  1          1   
2                           21.0         196                  1          1   
3                           29.0         196                  1          1   
4                           28.0         196                  1          1   

   Unnamed: 0_y product_name  aisle_id  department_id  prices _merge  
0           195         Soda        77              7     9.0   both  
1           195         Soda        77              7     9.0   both  
2           195         Soda        77              7     9.0   both  
3           195         Soda        77              7     9.0   both  
4           195         Soda        77              7     9.0   both

In [ ]:
# Drop first 3 (due to eval_set) unneccesary columns
df_merged = df_merged.drop(columns = ['Unnamed: 0.1','Unnamed: 0_x','eval_set'])

In [27]:
df_merged.head()

order_id  user_id  order_number  orders_day_of_week  order_hour_of_day  \
0   2539329        1             1                   2                  8   
1   2398795        1             2                   3                  7   
2    473747        1             3                   3                 12   
3   2254736        1             4                   4                  7   
4    431534        1             5                   4                 15   

   days_following_previous_order  product_id  add_to_cart_order  reordered  \
0                            NaN         196                  1          0   
1                           15.0         196                  1          1   
2                           21.0         196                  1          1   
3                           29.0         196                  1          1   
4                           28.0         196                  1          1   

   Unnamed: 0_y product_name  aisle_id  department_id  prices _merge  
0           195         Soda        77              7     9.0   both  
1           195         Soda        77              7     9.0   both  
2           195         Soda        77              7     9.0   both  
3           195         Soda        77              7     9.0   both  
4           195         Soda        77              7     9.0   both

*** Eval_set has been dropped again so we will see if it lasts this time, hopefully!

# Step 6 - Confirm results using merge flag

In [28]:
df_merged['_merge'].value_counts()

_merge
both          32434212
left_only            0
right_only           0
Name: count, dtype: int64

In [29]:
#checking merge shape
df_merged.shape

(32434212, 15)

This merge shape seems correct, as if the eval_set column may have also been dropped. (Hopefully, but will continue to check)

# Step 7 - Export!

In [30]:
#Export df_merged as pickle file b/c size
df_merged.to_pickle(os.path.join(path, '02 Data','Prepared Data', 'orders_products_merged.pkl'))